subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1662 @@
+		goto gc_done;
+	}
+
+	/* update sector_cycle */
+	fs->sector_cycle = empty_ate.cycle_cnt;
+
+	/* At this step empty & close ATEs are valid.
+	 * let's start the GC
+	 */
+	gc_addr &= ADDR_SECT_MASK;
+	gc_addr += close_ate.offset;
+
+	do {
+		gc_prev_addr = gc_addr;
+		rc = zms_prev_ate(fs, &gc_addr, &gc_ate);
+		if (rc) {
+			return rc;
+		}
+
+		if (!zms_ate_valid(fs, &gc_ate)) {
+			continue;
+		}
+
+#ifdef CONFIG_ZMS_LOOKUP_CACHE
+		wlk_addr = fs->lookup_cache[zms_lookup_cache_pos(gc_ate.id)];
+
+		if (wlk_addr == ZMS_LOOKUP_CACHE_NO_ADDR) {
+			wlk_addr = fs->ate_wra;
+		}
+#else
+		wlk_addr = fs->ate_wra;
+#endif
+		/* initialize the previous_sector_num to an impossible value
+		 * to read the sector cycle in the first loop
+		 */
+		previous_sector_num = -1;
+
+		do {
+			wlk_prev_addr = wlk_addr;
+			rc = zms_prev_ate(fs, &wlk_addr, &wlk_ate);
+			if (rc) {
+				return rc;
+			}
+			/* if ate with same id is reached we might need to copy.
+			 * only consider valid wlk_ate's. Something wrong might
+			 * have been written that has the same ate but is
+			 * invalid, don't consider these as a match.
+			 */
+			if (wlk_ate.id == gc_ate.id) {
+				/* read the ate cycle only when we change the sector
+				 * or if it is the first read
+				 */
+				rc = zms_get_cycle_on_sector_change(
+					fs, wlk_prev_addr, previous_sector_num, &current_cycle);
+				if (rc) {
+					return rc;
+				}
+				if (zms_ate_valid_different_sector(fs, &wlk_ate, current_cycle)) {
+					break;
+				}
+				previous_sector_num = SECTOR_NUM(wlk_prev_addr);
+			}
+		} while (wlk_addr != fs->ate_wra);
+
+		/* if walk has reached the same address as gc_addr copy is
+		 * needed unless it is a deleted item.
+		 */
+		if ((wlk_prev_addr == gc_prev_addr) && gc_ate.len) {
+			/* copy needed */
+			LOG_DBG("Moving %d, len %d", gc_ate.id, gc_ate.len);
+
+			if (gc_ate.len > 8) {
+				/* Copy Data only when len > 8
+				 * Otherwise, Data is already inside ATE
+				 */
+				data_addr = (gc_prev_addr & ADDR_SECT_MASK);
+				data_addr += gc_ate.offset;
+				gc_ate.offset = (uint16_t)(fs->data_wra & ADDR_OFFS_MASK);
+
+				rc = zms_flash_block_move(fs, data_addr, gc_ate.len);
+				if (rc) {
+					return rc;
+				}
+			}
+
+			gc_ate.cycle_cnt = previous_cycle;
+			zms_ate_crc8_update(&gc_ate);
+			rc = zms_flash_ate_wrt(fs, &gc_ate);
+			if (rc) {
+				return rc;
+			}
+		}
+	} while (gc_prev_addr != stop_addr);
+
+gc_done:
+
+	/* restore the previous sector_cycle */
+	fs->sector_cycle = previous_cycle;
+
+	/* Make it possible to detect that gc has finished by writing a
+	 * gc done ate to the sector. In the field we might have zms systems
+	 * that do not have sufficient space to add this ate, so for these
+	 * situations avoid adding the gc done ate.
+	 */
+
+	if (fs->ate_wra >= (fs->data_wra + ate_size)) {
+		rc = zms_add_gc_done_ate(fs);
+		if (rc) {
+			return rc;
+		}
+	}
+
+	/* Erase the GC'ed sector when needed */
+	if (ebw_required) {
```

Comment:
If you would move the `ebw_required` retrieval and check to `zms_flash_erase_sector` this would look better.